File: app/local_plugins/core/ai_plugin_interface/android/src/main/kotlin/com/aiagent/plugin_interface/NativeAgentRegistry.kt
package com.aiagent.plugin_interface

import android.util.Log

/**
 * NativeAgentRegistry — Agent 类型插件注册表（全局单例）
 *
 * 每个 Agent 类型插件在 FlutterPlugin.onAttachedToEngine() 时注册自己的工厂方法。
 * agents_server 通过 agentType 名称创建对应的 NativeAgent 实例。
 *
 * 使用示例：
 *   // 注册（agent_chat 插件的 onAttachedToEngine 中）
 *   NativeAgentRegistry.register("chat") { ChatAgentSession() }
 *
 *   // 创建（agents_server 收到 createAgent 命令时）
 *   val agent = NativeAgentRegistry.create("chat")
 *   agent.initialize(config, eventSink, context)
 */
object NativeAgentRegistry {

    private const val TAG = "NativeAgentRegistry"

    private val factories = mutableMapOf<String, () -> NativeAgent>()

    /**
     * 注册 Agent 类型工厂
     * @param agentType  类型标识（"chat", "sts", "translate", "ast"）
     * @param factory    创建 NativeAgent 实例的工厂方法
     */
    fun register(agentType: String, factory: () -> NativeAgent) {
        factories[agentType] = factory
        Log.d(TAG, "Registered agent type: $agentType")
    }

    /**
     * 创建指定类型的 NativeAgent 实例
     * @param agentType  类型标识
     * @return 新的 NativeAgent 实例（未初始化，需调用 initialize）
     */
    fun create(agentType: String): NativeAgent =
        factories[agentType]?.invoke()
            ?: throw IllegalArgumentException(
                "No agent registered for type: $agentType. Available: ${factories.keys}"
            )

    /** 查询所有已注册的 Agent 类型 */
    fun supportedTypes(): Set<String> = factories.keys
}

File: apps/client/local_plugins/core/ai_plugin_interface/ios/Classes/NativeAgentRegistry.swift
import Foundation
import os.log

/// Global registry of native agent factories.
///
/// Each agent-type plugin (`agent_chat`, `agent_sts_chat`, …) calls
/// `register(type:factory:)` during its iOS `FlutterPlugin.register(with:)`
/// hook. `agents_server` then resolves agent instances by type string.
public final class NativeAgentRegistry {
    public static let shared = NativeAgentRegistry()

    private let logger = OSLog(subsystem: "com.aiagent.plugin_interface", category: "NativeAgentRegistry")
    private let queue = DispatchQueue(label: "NativeAgentRegistry.factories")
    private var factories: [String: () -> NativeAgent] = [:]

    private init() {}

    /// Register an agent factory.
    /// - Parameters:
    ///   - agentType: canonical type string ("chat", "sts-chat", "translate", "ast-translate").
    ///   - factory: closure producing a fresh `NativeAgent` instance.
    public func register(_ agentType: String, factory: @escaping () -> NativeAgent) {
        queue.sync { factories[agentType] = factory }
        os_log("Registered agent type: %{public}@", log: logger, type: .debug, agentType)
    }

    /// Instantiate an agent. Throws if the type is unknown.
    public func create(_ agentType: String) throws -> NativeAgent {
        let factory = queue.sync { factories[agentType] }
        guard let factory = factory else {
            let known = queue.sync { Array(factories.keys) }
            throw NativeServiceError.invalidConfig(
                "No agent registered for type: \(agentType). Available: \(known)"
            )
        }
        return factory()
    }

    public func supportedTypes() -> Set<String> {
        Set(queue.sync { factories.keys })
    }
}
